src/stcal/dark_current/dark_sub.py

```diff
@@ -0,0 +1,391 @@
+    sci_groupgap = science_data.exp_groupgap
+
+    if instrument == 'MIRI':
+        drk_nints = dark_data.data.shape[0]
+        drk_ngroups = dark_data.data.shape[1]
+    else:
+        drk_nints = 1
+        drk_ngroups = dark_data.data.shape[0]
+
+    drk_nframes = dark_data.exp_nframes
+    drk_groupgap = dark_data.exp_groupgap
+
+    log.info(
+        'Science data nints=%d, ngroups=%d, nframes=%d, groupgap=%d',
+        sci_nints, sci_ngroups, sci_nframes, sci_groupgap
+    )
+    log.info(
+        'Dark data nints=%d, ngroups=%d, nframes=%d, groupgap=%d',
+        drk_nints, drk_ngroups, drk_nframes, drk_groupgap
+    )
+
+    # Check that the number of groups in the science data does not exceed
+    # the number of groups in the dark current array.
+    sci_total_frames = sci_ngroups * (sci_nframes + sci_groupgap)
+    drk_total_frames = drk_ngroups * (drk_nframes + drk_groupgap)
+    if sci_total_frames > drk_total_frames:
+        log.warning(
+            "Not enough data in dark reference file to match to "
+            "science data."
+        )
+        log.warning("Input will be returned without subtracting dark current.")
+        science_data.cal_step = 'SKIPPED'
+        out_data = copy.deepcopy(science_data)
+
+        return out_data, None
+
+    # Check that the value of nframes and groupgap in the dark
+    # are not greater than those of the science data
+    if drk_nframes > sci_nframes or drk_groupgap > sci_groupgap:
+        log.warning(
+            "The value of nframes or groupgap in the dark data is "
+            "greater than that of the science data."
+            "Input will be returned without subtracting dark current."
+        )
+        science_data.cal_step = 'SKIPPED'
+        out_data = copy.deepcopy(science_data)
+        return out_data, None
+
+    # Replace NaN's in the dark with zeros
+    dark_data.data[np.isnan(dark_data.data)] = 0.0
+
+    # Check whether the dark and science data have matching
+    # nframes and groupgap settings.
+    averaged_dark = None
+    if sci_nframes == drk_nframes and sci_groupgap == drk_groupgap:
+
+        # They match, so we can subtract the dark ref file data directly
+        output_data = subtract_dark(science_data, dark_data)
+
+        # If the user requested to have the dark file saved,
+        # save the reference model as this file. This will
+        # ensure consistency from the user's standpoint
+        if dark_output is not None:
+            averaged_dark = dark_data
+            averaged_dark.save = True
+            averaged_dark.output_name = dark_output
+
+    else:
+
+        # Create a frame-averaged version of the dark data to match
+        # the nframes and groupgap settings of the science data.
+        # If the data are from MIRI, the darks are integration-dependent and
+        # we average them with a seperate routine.
+
+        if instrument == 'MIRI':
+            averaged_dark = average_MIRIdark_frames(
```

Comment:
I wonder if it's possible to consolidate the `average_MIRIdark_frames` and `average_dark_frames` into a single function, which just checks the dimensions of the dark_data (i.e. if it's 4-D, then handle multiple integrations, otherwise ignore the integrations axis), rather than having hardwired MIRI-specific functions and code hooks.